Adds minikube support for pre-built images usage (#1)
accuknox/cilium-spire-tutorials: scenario01/README.md

## Scenario 1: L3/L4 policies based on SPIFFE ID

> 

Reviewer: ```suggestion

Description of the scenario: the goal of the scenario exposed by the image below is to do a connectivity test (icmp) between the pods `poddefault` and `podfoo`. A Cilium Network Policy (CNP) is going to be applied to allow the connectivity between both pods based on the SPIFFE ID label received by the pod. For this tutorial the following steps will be performed:

1. Create all pods for this scenario;
2. Do the connectivity test between them before apply a CNP. The connectivity test should worked normally;
3. Deny all the communication between the pods using a CNP. Do a connectivity test between them to see if fails (as expected);
4. Allow the communication between both pods and do a connectivity again - should worked;
5. From `poddefault`, try to ping outside (8.8.8.8) - should be blocked.
```

<img src="../imgs/scenario01.png" alt="drawing" width="800"/>

> Create spire registration entries

Reviewer: ```suggestion
Create registration entries in Spire. The registration entries contains the SPIFFE ID based on a set of selectors, which is going to be assigned to the pods. Run the follow to create all registration entries for this tutorial after 
```

```

> ./0-create-entries.sh

Reviewer: ```suggestion
% ./0-create-entries.sh
```

```

Deploy pods `poddefault` and `podfoo`. The former is deployed on namespace `default`. The other one is deployed on namespace `foo`.

```

> kubectl apply -f 1-http-sw-app.yaml

Reviewer: ```suggestion
% kubectl apply -f 1-http-sw-app.yaml
```

```

> Check if the SPIFFE ID was assigned in `swing` and `deathstar` pods:

Reviewer: ```suggestion
Check if the SPIFFE ID was assigned to `poddefault` and `podfoo` pods. The pod `poddefault` must have a new label called `spiffe://example.org/ns/default/sa/default` and the pod `podfoo` a new label called `spiffe://example.org/ns/default/sa/foo`.
```

> Check if the SPIFFE ID was assigned in `swing` and `deathstar` pods:

Reviewer: ```suggestion
Check if the SPIFFE ID was assigned in `xwing` and `deathstar` pods:
```

```

> kubectl -n kube-system exec cilium-74m7n -- cilium endpoint list

Reviewer: ```suggestion
% kubectl -n kube-system exec cilium-74m7n -- cilium endpoint list
```

                                                           k8s:type=my-pod                                                                                       
                                                           spiffe://example.org/ns/default/sa/default                                                            
3362       Disabled           Disabled          12147      k8s:app=spire-server                                                             10.0.0.140   ready   
                                                           k8s:io.cilium.k8s.policy.cluster=default                                                              
                                                           k8s:io.cilium.k8s.policy.serviceaccount=spire-server                                                  
                                                           k8s:io.kubernetes.pod.namespace=spire                                                                 
                                                           k8s:statefulset.kubernetes.io/pod-name=spire-server-0     
```

> Lets deny all the  communication only between them. 

Reviewer: ```suggestion
Do the connectivity test between them before apply a CNP - it should worked. Now, lets deny all the  communication only between them: 
```

```

> kubectl apply -f 2-deny-all.yaml

Reviewer: ```suggestion
% kubectl apply -f 2-deny-all.yaml
```

```

Check the if the communication will be blocked between the endpoints:

```

> kubectl exec poddefault -- ping <ip-podfoo>

Reviewer: ```suggestion
% kubectl exec poddefault -- ping <ip-podfoo>
```

```

> The podfoo IP address is found using the command `kubectl get pods poddefault -o wide`

> Lets allow the communication from poddefault to podfoo. Note the policy below is based on the SPIFFE ID that were assigned in each endpoint.

Reviewer: ```suggestion
Lets allow the communication from `poddefault` to `podfoo`. Note the policy below is based on the SPIFFE ID that were assigned to each endpoint.
```

```
# Allow pods running with the default service account to contact pods
# running with the foo service account.
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "spiffe-based"
spec:
  endpointSelector:
    matchLabels:
      spiffe://example.org/ns/default/sa/default: ""
  egress:
  - toEndpoints:
    - matchLabels:
        spiffe://example.org/ns/default/sa/foo: ""

```

> Apply Cilium Network Policies (CNPs):

Reviewer: ```suggestion
Apply Cilium Network Policies (CNP):
```

```
kubectl apply -f 3-spiffe-based.yaml
```

> Let's check the if the communication is allowed between the endpoints with this new policy:

Reviewer: ```suggestion
Lets check the if the communication is allowed between the endpoints using this new policy:
```

```

> kubectl exec poddefault -- ping <ip-podfoo>

Reviewer: ```suggestion
% kubectl exec poddefault -- ping <ip-podfoo>
```

PING 10.0.0.84 (10.0.0.84) 56(84) bytes of data.
64 bytes from 10.0.0.84: icmp_seq=1 ttl=64 time=0.059 ms
64 bytes from 10.0.0.84: icmp_seq=2 ttl=64 time=0.054 ms
```

Great! The policy enforcement based on SPIFFE ID label is working.

> 

Reviewer: ```suggestion
If the connectivity is performed to outside (8.8.8.8), it should not be allowed.

% kubectl exec poddefault -- ping 8.8.8.8

```

To clean up the pods and CNPs:

```

> ./4-clean-all.sh

Reviewer: ```suggestion
% ./4-clean-all.sh
```
